### PR TITLE
Convert checkout API to use tokens instead of full objects

### LINKED ISSUE #200: Convert checkout APIs to accept a token instead of complete object
Currently, some checkout API unnecessarily require a `BUYCheckout` object when in reality, only the `token` is required.

### PR DESCRIPTION
**Pull request by dbart01**
### What this does
- updates checkout API to only require a checkout `token` instead of `BUYCheckout`
- updates `BUYCompleteCheckout` operation to require only the checkout `token`

Fixes #200 

@bgulanowski @davidmuzi

### CODE REVIEW

**Comment by dbart01** [comment]
Ready for review.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.h [hunk 1]
```diff
@@ -92,81 +94,56 @@ typedef void (^BUYDataGiftCardBlock)(BUYGiftCard * _Nullable giftCard, NSError *
...
- *
- *  @param checkout The BUYCheckout to updated on Shopify
- *  @param block    (^BUYDataCheckoutBlock)(BUYCheckout *checkout, NSError *error);
+ *  @param checkoutToken The checkout token for which to retrieve the updated checkout
+ *  @param block         (^BUYDataCheckoutBlock)(BUYCheckout *checkout, NSError *error);
  *
  *  @return The associated BUYRequestOperation
  */
-- (BUYRequestOperation *)updateCheckout:(BUYCheckout *)checkout completion:(BUYDataCheckoutBlock)block;
+- (BUYRequestOperation *)getCheckoutWithToken:(NSString *)checkoutToken completion:(BUYDataCheckoutBlock)block;
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.h]: Nitpick: it's a bit subtle, but we should prefer "For" instead of "With" (the preposition) in most of these method signatures. "With" is best reserved for initialization arguments.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.h [hunk 2]
```diff
@@ -212,16 +189,40 @@ typedef void (^BUYDataGiftCardBlock)(BUYGiftCard * _Nullable giftCard, NSError *
...
 /**
+ *  Applies a gift card code to the checkout.
+ *
+ *  @param giftCardCode The gift card code to apply on an existing checkout on Shopify. Note: This is not the same as the gift card identifier.
+ *  @param checkout     An existing BUYCheckout on Shopify
+ *  @param block        (^BUYDataCheckoutBlock)(BUYCheckout *checkout, NSError *error);
+ *
+ *  @return The associated BUYRequestOperation
+ */
+- (BUYRequestOperation *)applyGiftCardWithCode:(NSString *)giftCardCode toCheckout:(BUYCheckout *)checkout completion:(BUYDataCheckoutBlock)block;
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.h]: Remove the "With" here. Just "applyGiftCardCode". I know this is the original, but this signature is too wordy.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.h [hunk 3]
```diff
@@ -92,81 +94,56 @@ typedef void (^BUYDataGiftCardBlock)(BUYGiftCard * _Nullable giftCard, NSError *
...
  *
- *  @param checkout The BUYCheckout to retrieve completion status for
- *  @param block    (^BUYDataStatusBlock)(BUYCheckout *checkout, BUYStatus status, NSError *error);
+ *  @param checkoutToken  The checkout token for which to retrieve completion status
+ *  @param block          (^BUYDataStatusBlock)(BUYCheckout *checkout, BUYStatus status, NSError *error);
  *
  *  @return The associated BUYRequestOperation
  */
-- (BUYRequestOperation *)getCompletionStatusOfCheckout:(BUYCheckout *)checkout completion:(BUYDataStatusBlock)block;
+- (BUYRequestOperation *)getCompletionStatusOfCheckoutWithToken:(NSString *)checkoutToken completion:(BUYDataStatusBlock)block;
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.h]: In this signature, the second preposition, "With", is not needed. `-getCompletionStatusForCheckoutToken:` would be sufficient.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m [hunk 4]
```diff
@@ -205,26 +119,81 @@ - (BUYRequestOperation *)getCompletionStatusOfCheckoutURL:(NSURL *)url completio
...
+- (void)handleCheckoutResponse:(NSDictionary *)json error:(NSError *)error block:(BUYDataCheckoutBlock)block
+{
+	BUYCheckout *checkout = nil;
+	if (json && !error) {
+		checkout = [self.modelManager insertCheckoutWithJSONDictionary:json[@"checkout"]];
+	}
+	block(checkout, error);
+}
+
+- (void)updateCheckout:(BUYCheckout *)checkout withGiftCardDictionary:(NSDictionary *)giftCardDictionary addingGiftCard:(BOOL)addingGiftCard
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m]: I think we should break up this method. We can extract the identical logic to another method, but it's only two lines.

Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m]: Actually, we should merge this method back into the calling methods. It's only called from two places. It's a confusing method.

File: Mobile Buy SDK/Mobile Buy SDK Tests/BUYIntegrationTest.m [hunk 5]
```diff
@@ -1025,7 +1021,7 @@ - (void)testExpiringCheckout
 	
 	// Expire the checkout
 	XCTestExpectation *expectation2 = [self expectationWithDescription:NSStringFromSelector(_cmd)];
-	[self.client removeProductReservationsFromCheckout:_checkout completion:^(BUYCheckout *returnedCheckout, NSError *error) {
+	[self.client removeProductReservationsFromCheckoutForToken:_checkout.token completion:^(BUYCheckout *returnedCheckout, NSError *error) {
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK Tests/BUYIntegrationTest.m]: `removeProductReservationsForCheckoutToken:` would be more concise.

Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK Tests/BUYIntegrationTest.m]: Actually that would be inconsistent with other method names, so this is fine.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m [hunk 6]
```diff
@@ -266,11 +218,68 @@ - (BUYRequestOperation *)storeCreditCard:(BUYCreditCard *)creditCard checkout:(B
...
+	NSURL *route = [self urlForCheckoutsUsingGiftCardForToken:checkout.token];
+	
+	return [self postRequestForURL:route object:giftCard completionHandler:^(NSDictionary *json, NSHTTPURLResponse *response, NSError *error) {
+		if (json && !error) {
+			
+			NSDictionary *giftCardJSON = json[@"gift_card"];
+			BUYGiftCard *giftCard      = [self.modelManager insertGiftCardWithJSONDictionary:giftCardJSON];
+			
+			/* ---------------------------------
+			 * We have to update a copy of the
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m]: This comment is not consistent with the code.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Routing.h [hunk 7]
```diff
@@ -36,13 +36,13 @@
 - (NSURL *)urlForCollectionListingsWithParameters:(NSDictionary *)parameters;
 
 - (NSURL *)urlForCheckouts;
-- (NSURL *)urlForCheckoutsWithToken:(NSString *)token;
-- (NSURL *)urlForCheckoutsProcessingWithToken:(NSString *)token;
-- (NSURL *)urlForCheckoutsCompletionWithToken:(NSString *)token;
-- (NSURL *)urlForCheckoutsShippingRatesWithToken:(NSString *)token parameters:(NSDictionary *)parameters;
+- (NSURL *)urlForCheckoutsForToken:(NSString *)token;
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Routing.h]: These method signatures sound bizarre. (They did before, too.) Is this functionality private? I don't want to get hung up on this, but would like to be able to change this later. Would really like `-urlForCheckoutToken:` etc. but don't know if that's inaccurate.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Address.h [hunk 8]
```diff
@@ -98,13 +98,13 @@ typedef void (^BUYDataAddressBlock)(BUYAddress * _Nullable address, NSError * _N
...
  *
- *  @param address    Address to delete
+ *  @param addressID  Address ID to delete
  *  @param customerID Customer ID for which to delete the address
  *  @param block      (BUYStatus status, NSError *error)
  *
  *  @return The associated BUYRequestOperation
  */
-- (BUYRequestOperation *)deleteAddress:(BUYAddress *)address forCustomerID:(NSString *)customerID callback:(BUYDataStatusBlock)block;
+- (BUYRequestOperation *)deleteAddressWithID:(NSNumber *)addressID forCustomerID:(NSString *)customerID callback:(BUYDataStatusBlock)block;
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Address.h]: Nitpick: can we drop the "for" in front of "Customer"? `-deleteAddressWithID:customerID:callback:`

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m (hunk 6 above)
Comment (dbart01) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Checkout.m]: This is addressed in #204, I'd rather not duplicate the fix.

File: Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Routing.h (hunk 7 above)
Comment (dbart01) [inline Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient+Routing.h]: Agreed. Would be nice to make this more natural sounding, but this is private and we can do so at any time in the future.

**Comment by bgulanowski** [comment]
👍 :shipit:
